File: modules/myoblue-native/src/MyoblueNativeModule.ts
import { NativeModule, requireOptionalNativeModule } from "expo";

export type SensorState = "connecting" | "discovering" | "live" | "reconnecting" | "disconnected" | "failed";

export type MyoblueEvents = {
  onDevice: (e: { id: string; name: string; rssi: number }) => void;
  onSensorState: (e: { id: string; name: string; state: SensorState }) => void;
  /** data: base64 of the 244-byte packet; t: ms on the elapsedRealtime clock */
  onPacket: (e: { id: string; data: string; t: number }) => void;
  onMarker: (e: { t: number; label: string }) => void;
  onRecording: (e: { active: boolean }) => void;
  /** self-update progress: download 0..1, install, confirm (system prompt shown), done, error */
  onUpdate: (e: { phase: "download" | "install" | "confirm" | "done" | "error"; progress: number; message: string | null }) => void;
};

export interface RecordingStatus {
  path: string;
  elapsedMs: number;
  startedAt: number;
  packets: number;
  sensors: number;
  markers: number;
}

declare class MyoblueNativeModule extends NativeModule<MyoblueEvents> {
  isBluetoothOn(): boolean;
  startScan(): boolean;
  stopScan(): void;
  connect(id: string, name: string): void;
  disconnect(id: string): void;
  sensors(): { id: string; name: string; state: SensorState; packets: number }[];
  now(): number;
  injectPacket(id: string, name: string, base64: string): void;
  startRecording(path: string, title: string): boolean;
  stopRecording(): { path: string; durationMs: number; packets: number } | null;
  addMarker(label: string): number;
  recordingStatus(): RecordingStatus | null;
  ignoringBatteryOptimizations(): boolean;
  openBatterySettings(): void;
  appVersionCode(): number;
  /** copy a file into Downloads/MyoLift (visible over USB); returns the path shown to the user */
  saveToDownloads(src: string, name: string): Promise<string>;
  canInstallUpdates(): boolean;
  openInstallPermission(): void;
  /** download the APK and hand it to the system installer; false if an update is already running */
  installUpdate(url: string): boolean;
}

/** null when the native module isn't compiled in (e.g. web or Expo Go). */
export default requireOptionalNativeModule<MyoblueNativeModule>("MyoblueNative");

File: modules/myoblue-native/android/src/main/java/expo/modules/myobluenative/MyoblueNativeModule.kt
package expo.modules.myobluenative

import android.util.Base64
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition

/** JS bridge for MyoBle (Bluetooth + recording). Packets are sent to JS base64-encoded. */
class MyoblueNativeModule : Module() {
  override fun definition() = ModuleDefinition {
    Name("MyoblueNative")

    Events("onDevice", "onSensorState", "onPacket", "onMarker", "onRecording")

    OnCreate {
      val context = appContext.reactContext ?: return@OnCreate
      MyoBle.init(context)
      MyoBle.listener = object : MyoBle.Listener {
        override fun onDevice(id: String, name: String, rssi: Int) =
          sendEvent("onDevice", mapOf("id" to id, "name" to name, "rssi" to rssi))
        override fun onSensorState(id: String, name: String, state: String) =
          sendEvent("onSensorState", mapOf("id" to id, "name" to name, "state" to state))
        override fun onPacket(id: String, data: ByteArray, t: Double) =
          sendEvent("onPacket", mapOf("id" to id, "data" to Base64.encodeToString(data, Base64.NO_WRAP), "t" to t))
        override fun onMarker(t: Double, label: String) =
          sendEvent("onMarker", mapOf("t" to t, "label" to label))
        override fun onRecording(active: Boolean) =
          sendEvent("onRecording", mapOf("active" to active))
      }
    }

    OnDestroy { MyoBle.listener = null }

    Function("isBluetoothOn") { MyoBle.isEnabled() }
    Function("startScan") { MyoBle.startScan() }
    Function("stopScan") { MyoBle.stopScan() }
    Function("connect") { id: String, name: String -> MyoBle.connect(id, name) }
    Function("disconnect") { id: String -> MyoBle.disconnect(id) }
    Function("sensors") { MyoBle.sensors() }
    Function("now") { MyoBle.now() }
    Function("injectPacket") { id: String, name: String, data: String ->
      MyoBle.inject(id, name, Base64.decode(data, Base64.NO_WRAP))
    }

    Function("startRecording") { path: String, title: String -> MyoBle.Recorder.start(path, title) }
    Function("stopRecording") { MyoBle.Recorder.stop() }
    Function("addMarker") { label: String -> MyoBle.Recorder.marker(label) }
    Function("recordingStatus") { MyoBle.Recorder.status() }

    Function("ignoringBatteryOptimizations") {
      val context = appContext.reactContext ?: return@Function true
      MyoBle.ignoringBatteryOptimizations(context)
    }
    Function("openBatterySettings") {
      val context = appContext.reactContext
      if (context != null) MyoBle.openBatterySettings(context)
    }
  }
}
